### PR TITLE
feat(samtools): Update

### PR DESCRIPTION
**Pull request by henrikstranneheim**
This PR adds/fixes:

- Update samtools
- Refactor code

**Review:**
- [x] Tests executed by HS

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: crunchy/cli/compress_cmd.py [hunk 1]
```diff
@@ -78,10 +76,7 @@ def fastq(
         first_read=first_read, second_read=second_read, spring=spring_path
     )
 
-    metadata_path = None
-    if metadata_file:
-        metadata_path = dump_spring_metadata(metadata)
-
+    metadata_path = dump_spring_metadata(metadata) if metadata_file else None
```
Comment (karlnyr) [inline crunchy/cli/compress_cmd.py:79]:
```suggestion
    metadata_path: Optional[Path] = dump_spring_metadata(metadata) if metadata_file else None
```

File: crunchy/cli/compress_cmd.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-"""CLI functions to compress"""
+"""CLI functions to compress."""
 import logging
 import pathlib
```
Comment (karlnyr) [inline crunchy/cli/compress_cmd.py:4]:
```suggestion
from typing import Optional
```

File: crunchy/cli/compress_cmd.py [hunk 3]
```diff
@@ -146,27 +141,25 @@ def fastq(
...
     if dry_run:
         LOG.info("Dry Run! No files will be created or deleted")
     cram_api = ctx.obj.get("cram_api")
     try:
         cram_api.self_check()
-    except (SyntaxError, FileNotFoundError):
-        raise click.Abort
+    except (SyntaxError, FileNotFoundError) as error:
+        raise click.Abort from error
     bam_path = pathlib.Path(bam_path)
```
Comment (karlnyr) [inline crunchy/cli/compress_cmd.py:153]:
```suggestion
    bam_path: pathlib.Path = pathlib.Path(bam_path)
```

File: crunchy/cli/compress_cmd.py [hunk 4]
```diff
@@ -146,27 +141,25 @@ def fastq(
...
-    if not cram_path:
-        cram_path = cram_outpath(bam_path)
-    else:
-        cram_path = pathlib.Path(cram_path)
+    cram_path = pathlib.Path(cram_path) if cram_path else cram_outpath(bam_path)
```
Comment (karlnyr) [inline crunchy/cli/compress_cmd.py:154]:
```suggestion
    cram_path: pathlib.Path = pathlib.Path(cram_path) if cram_path else cram_outpath(bam_path)
```

File: crunchy/cli/compress_cmd.py [hunk 5]
```diff
@@ -146,27 +141,25 @@ def fastq(
 @click.option("--dry-run", is_flag=True)
 @click.pass_context
 def bam(ctx, bam_path, cram_path, dry_run):
```
Comment (karlnyr) [inline crunchy/cli/compress_cmd.py:143]:
```suggestion
def bam(ctx, bam_path: click.Path, cram_path: str, dry_run: bool):
```

File: crunchy/command.py [hunk 6]
```diff
@@ -87,23 +86,21 @@ def stderr(self):
...
 
     def __repr__(self):
         return f"Process:base_call:{self.base_call}"
 
 
 class SpringProcess(Process):
-    """Process to deal with spring commands."""
+    """Process to deal with Sring commands."""
 
     def __init__(self, binary, threads=8, tmp_dir=None):
```
Comment (karlnyr) [inline crunchy/command.py:102]:
```suggestion
    def __init__(self, binary: str, threads: int = 8, tmp_dir: Optional[str] = None):
```

File: crunchy/command.py [hunk 7]
```diff
@@ -87,23 +86,21 @@ def stderr(self):
...
-        """Initialise a spring process"""
+        """Initialise a spring process."""
         super().__init__(binary)
         self.threads = threads
         self.tmp = tmp_dir
```
Comment (karlnyr) [inline crunchy/command.py:106]:
```suggestion
        self.threads: int = threads
        self.tmp: Optional[str] = tmp_dir
```

File: crunchy/command.py [hunk 8]
```diff
@@ -162,7 +159,7 @@ def compress(
         if self.tmp is not None:
```
Comment (karlnyr) [inline crunchy/command.py:159]:
```suggestion
        if self.tmp:
```

File: crunchy/command.py [hunk 9]
```diff
@@ -187,17 +184,17 @@ def __repr__(self):
 
 
 class CramProcess(Process):
-    """Process to deal with cram commands."""
+    """Process to deal with CRAM commands."""
 
     def __init__(self, binary: str, refgenome_path: str, threads=8):
```
Comment (karlnyr) [inline crunchy/command.py:189]:
```suggestion
    def __init__(self, binary: str, refgenome_path: str, threads: int = 8):
```

File: crunchy/command.py [hunk 10]
```diff
@@ -187,17 +184,17 @@ def __repr__(self):
...
-        """Initialise a spring process"""
+        """Initialise a spring process."""
         super().__init__(binary)
         self.refgenome_path = refgenome_path
         self.threads = threads
```
Comment (karlnyr) [inline crunchy/command.py:193]:
```suggestion
        self.refgenome_path: str = refgenome_path
        self.threads: int = threads
```

**Review by karlnyr — COMMENTED** [review]
👍

**Comment by henrikstranneheim** [comment]
@karlnyr Thx for the review. Fixed all comments above

File: crunchy/command.py [hunk 11]
```diff
@@ -19,19 +19,19 @@ class Process:
     called, that will be handled in this module.Output form stdout and stdin will be handeld here.
     """
 
-    def __init__(self, binary, config=None, config_parameter="--config"):
+    def __init__(self, binary: str, config: str = None, config_parameter: str ="--config"):
```
Comment (karlnyr) [inline crunchy/command.py:22]:
```suggestion
    def __init__(self, binary: str, config: Optinal[str] = None, config_parameter: str ="--config"):
```

File: crunchy/command.py [hunk 12]
```diff
@@ -1,13 +1,13 @@
...
+Code to handle communications to the shell.
 """
 
 import copy
 import logging
-import pathlib
 import subprocess
 from pathlib import Path
 from subprocess import CalledProcessError
+from typing import Optional
```
Comment (karlnyr) [inline crunchy/command.py:10]:
```suggestion
from typing import Iterator, List, Optional
```

File: crunchy/command.py [hunk 13]
```diff
@@ -19,19 +19,19 @@ class Process:
...
 
-    def __init__(self, binary, config=None, config_parameter="--config"):
+    def __init__(self, binary: str, config: str = None, config_parameter: str ="--config"):
         """
         Args:
             binary(str): Path to binary for the process to use
             config(str): Path to config if used by process
         """
         super(Process, self).__init__()
         self.binary = binary
```
Comment (karlnyr) [inline crunchy/command.py:29]:
```suggestion
        self.binary: str = binary
```

File: crunchy/command.py [hunk 14]
```diff
@@ -19,19 +19,19 @@ class Process:
...
-        LOG.info("Initialising Process with binary: %s", self.binary)
+        LOG.info(f"Initialising Process with binary: {self.binary}")
         self.base_call = [self.binary]
```
Comment (karlnyr) [inline crunchy/command.py:31]:
```suggestion
        self.base_call: List[str] = [self.binary]
```

File: crunchy/command.py [hunk 15]
```diff
@@ -19,19 +19,19 @@ class Process:
...
         if config:
             self.base_call.extend([config_parameter, config])
-        LOG.info("Use base call %s", self.base_call)
+        LOG.info(f"Use base call {self.base_call}")
         self._stdout = ""
         self._stderr = ""
```
Comment (karlnyr) [inline crunchy/command.py:36]:
```suggestion
        self._stdout: str = ""
        self._stderr: str= ""
```

File: crunchy/command.py [hunk 16]
```diff
@@ -87,26 +87,24 @@ def stderr(self):
 
     def stdout_lines(self):
```
Comment (karlnyr) [inline crunchy/command.py:88]:
```suggestion
    def stdout_lines(self) -> Iterator[str]:
```

File: crunchy/command.py [hunk 17]
```diff
@@ -87,26 +87,24 @@ def stderr(self):
...
         """Iterate over the lines in self.stdout."""
-        for line in self.stdout.split("\n"):
-            yield line
+        yield from self.stdout.split("\n")
 
     def stderr_lines(self):
```
Comment (karlnyr) [inline crunchy/command.py:92]:
```suggestion
    def stderr_lines(self) -> Iterator[str]:
```

File: crunchy/command.py [hunk 18]
```diff
@@ -87,26 +87,24 @@ def stderr(self):
...
         """Iterate over the lines in self.stderr."""
-        for line in self.stderr.split("\n"):
-            yield line
+        yield from self.stderr.split("\n")
 
     def __repr__(self):
```
Comment (karlnyr) [inline crunchy/command.py:96]:
```suggestion
    def __repr__(self) -> str:
```

**Review by karlnyr — APPROVED** [review]

**Pull request merged**